Serial server stability

chrisguikema opened the pull request:
This PR adds a few commits that stabilize the serial server based on issues that DornerWorks has seen the last few years of development

Test with: seL4/camkes-vm#40

axel-h commented on components/SerialServer/src/serial.c:
```diff
@@ -507,10 +529,14 @@ void pre_init(void)
         for (int i = 0; i < num_getchar_clients; i++) {
             unsigned int badge = getchar_enumerate_badge(i);
-            assert(badge <= num_getchar_clients);
+            assert(badge <= 9);
```
Isn't the badge ID decoupled from the client ID now, so arbitrary badge IDs can be accepted here. Just the number of clients is limited to 10. Then this assert should be removed and the loop should become `for (int i = 0; i < MIN(num_getchar_clients, 10); i++) { ...`.

axel-h replied:
Ah no, ignore comment. But contrary to the commit comment, we are not using the badge as the index here. I'd say we want to keep the `assert(badge <= num_getchar_clients);` assuming badges are [0 to n-1] for n clients. We we need is something like:
```
c = &getchar_clients[i];
if (badge > 9) { 
    print("sorry, can use this client as badge %d exceeds range 0-9", badge); 
    c->idx = -1;
    continue;
}
c->idx = i; /* can use this one */
...
```
here.

chrisguikema replied:
Sorry, I had two commits that were doing similar things, so I merged them into one. I can update the commit message to better reflect the contents of the code change.

axel-h commented on components/SerialServer/src/serial.c:
```diff
@@ -393,23 +406,27 @@ static void handle_char(uint8_t c)
             statemachine = 1;
-            if (c >= '0' && c < '0' + num_getchar_clients) {
+            if (c >= '0' && c <= '9') {
                 int client = c - '0';
```
This should become `int badge = c - '0';` now, because we are selecting by badge ID.

axel-h commented on components/SerialServer/src/serial.c:
```diff
@@ -507,10 +526,18 @@ void pre_init(void)
             getchar_clients[i].buf = getchar_buf(badge);
             getchar_clients[i].last_head = -1;
+            if (i == 0) {
+                active_client = badge;
```
There is no guarantee that `active_client` is set here, because the `if (badge > MAX_BADGE) {` may trigger for the first loop iteration already. So `active_client` might be undefined - or well, the global variable is set to `0`. Might be better so initialize this with `-1`, as client[0] could be invalid.

axel-h replied:
Seem "-1" is used for multi-client already, so "-2" should be used then to indicate nothing is active.

chrisguikema replied:
Good idea. I initialized `active_client` to -2, and then checked against `-2` instead of `i==0`.

axel-h commented on components/SerialServer/src/serial.c:
```diff
@@ -310,9 +311,21 @@ static void internal_putchar(int b, int c)
 {
-    getchar_client_t *client = &getchar_clients[id];
+    int idx = getchar_index_from_badge(id);
+    ZF_LOGF_IF(idx == -1, "Badge %d is not valid\n", id);
```
I don't think this case is really fatal. We have the limitation of 10 clients now, so this client should simply be ignored then.
```
if (idx == -1) {
    ZF_LOGW("Client with badge %d is not supported\n", id);
    return
}
```

chrisguikema replied:
Done.

kent-mcleod commented on components/SerialServer/src/serial.c:
```diff
@@ -360,7 +360,9 @@ static void handle_char(uint8_t c)
     case 1:
-        if (c == ESCAPE_CHAR) {
+        if (c == '\r' || c == '\n') {
+            give_client_char(c);
```
Oh this should be a nice quality of life improvement.

kent-mcleod commented on components/SerialServer/src/serial.c:
```diff
@@ -404,16 +404,16 @@ static void handle_char(uint8_t c)
-                statemachine = 0;
-                give_client_char(ESCAPE_CHAR);
-                give_client_char(c);
+                printf(COLOR_RESET "\r\nInvalid SerialServer command: %c"
+                       "\r\nType %c? for help"
+                       "\r\n", (char)c, ESCAPE_CHAR);
```
If someone was trying to type `@foo` into a client, then the previous behavior would be that the client would see `@foo`.  The new behavior appears that the client will now see `oo` right? 

If a user now wanted the client to see `@foo` then they would need to input `@@foo`.  Does it make sense that this should be the only way to send strings starting with the `ESCAPE_CHAR` to the client? This would be safer for extending the serial server commands in the future as it would reserve all characters for possible serial server commands.

chrisguikema replied:
Correct. You'd have to double-type the `ESCAPE_CHAR` (Only after the user has entered a newline). I think the benefits outweigh the drawbacks, considering its an edge-case.

kent-mcleod commented on components/SerialServer/src/serial.c:
```diff
@@ -70,13 +72,11 @@ const char *all_output_colours[2][MAX_CLIENTS] = {
     {
         /* Processed streams */
         ANSI_COLOR(RED),
-        ANSI_COLOR(GREEN),
```
From the commit message: "and the "Green" color is removed from the Serial Server, as it is very difficult to read."
The rendered color for an ANSI color is affected by whatever local terminal emulator is being used (https://en.wikipedia.org/wiki/ANSI_escape_code#Colors). My opinion is that this entry shouldn't be removed for this reason. What do others think?

chrisguikema replied:
If the Max clients is dropped to 10, we'd have to remove one color. Personal preference is Green because `screen` and TeraTerm don't render it well, but I could always modify the PR/commit message based on community feedback.

kent-mcleod replied:
I don't understand why "A side effect of this change is that the SerialServer is now limited to 10 clients" is required?

kent-mcleod replied:
Ok I see.

kent-mcleod commented on components/SerialServer/src/serial.c:
```diff
@@ -24,7 +24,8 @@
 #define ESCAPE_CHAR '@'
-#define MAX_CLIENTS 12
+#define MAX_CLIENTS 10
+#define MAX_BADGE   (MAX_CLIENTS-1)
```
A badge value of 0 is actually the same as unbadged and so shouldn't be used as the client could make copies of the cap and give them any badge.

chrisguikema replied:
That would cause issues with this PR: https://github.com/seL4/camkes-vm/pull/40

How do you suggest we configure the badges/client relationship then?

kent-mcleod replied:
Currently, there's a different ID assignment for each of the interfaces, getchar, putchar and batch.  Making this ID assignment unified requires either making a single interface or making the system designer assign IDs across all of the connection instances manually. It used to be the second way but it was easy to mis-configure and led to lots of boilerplate. 

Maybe you just want to be able to assign a client_id to the getchar interface and don't change the ID assignment for the putchar/batch interfaces.

axel-h commented on components/SerialServer/src/serial.c:
```diff
@@ -45,6 +49,7 @@ typedef struct getchar_buffer {
 compile_time_assert(getchar_buf_sized, sizeof(client_buffer_t) == sizeof(Buf));
 
 typedef struct getchar_client {
+    unsigned int idx;
```
I think `idx` is not needed and make things more complicated than necessary. The field`client_id` could simply become `batch`. The helper `getchar_index_from_badge()` would become `getchar_client_from_badge()`

```
static getchar_client_t getchar_clients[MAX_BADGE];

getchar_client_t *getchar_client_from_badge(int badge)
{
    for (int i = 0; i < ARRAY_ELEMENTS(getchar_clients); i++) {
        if (getchar_clients[i].badge == badge) {
            return &getchar_clients[i];
        }
    }
    return null;
}
 
void pre_init(void)
{
    [...]

    int j = 0
    for (int i = 0; i < num_getchar_clients; i++) {
            unsigned int badge = getchar_enumerate_badge(i);
            if (badge > MAX_BADGE) {
                ZF_LOGE("Client badge %d exceeds range 0->%d", badge, MAX_BADGE);
                continue;
            }
            getchar_client_t *c = &getchar_clients[j++]
            c->badge = badge;
            c->buf = getchar_buf(badge);
            c->last_head = -1;
        
    [...]
```

chrisguikema replied:
I added your getchar_client_from_badge file. That's a good way to remove the index.

chrisguikema commented:
@kent-mcleod @axel-h I think I addressed all the comments here.

The big thing we're trying to allowing them to be nonsequential. I want to say we had a system where 1 VM was using network as a console instead of serial, and it completely messed up the serial server.

axel-h reviewed (approved):
Looks good to me

kent-mcleod commented:
> The big thing we're trying to allowing them to be nonsequential. I want to say we had a system where 1 VM was using network as a console instead of serial, and it completely messed up the serial server.

I guess I might still be missing what the underlying issue is? 
The serial server uses identifiers on both the tx and rx paths. On tx, the identifier is used to select a color and on the rx the identifier is used to select the client that the characters are delivered to (or to broadcast to all clients). IDs are currently assigned by the camkes connection enumeration order, which is kind of arbitrary. There's usually more tx clients than rx clients, because any component can putchar, but usually only VM components want to getchar. If you have 1 guest that's not using the serial server, then the serial server won't assign a rx ID for it.

Is the goal of this PR was to be able to match the rx client ID with the VM guest ID? So if you have a VM guest ID of 9, you get a client ID of 9 even if there are no clients in-between?

chrisguikema commented:
> Is the goal of this PR was to be able to match the rx client ID with the VM guest ID? So if you have a VM guest ID of 9, you get a client ID of 9 even if there are no clients in-between?

I believe that was our use-case, yes. I'd also be okay with splitting the PRs into the "stability fixes", and this "getchar refactor", if you'd like to still have a discussion on this part specifically.

chrisguikema commented:
Also, I want to say we had an issue with the getchar badge being randomly assigned from build to build, so you could end up with VMs having different client ids on a build to build basis. So typing "@0" would get you to VM2 sometimes, etc.

chrisguikema commented:
@kent-mcleod is this good to merge?

chrisguikema commented:
@kent-mcleod I split this up into the stability fixes, and once these are merged, I'll push up a branch that handles the client ID value of 0 properly.
